pkg/util/reflector/prometheus/prometheus.go

```diff
@@ -0,0 +1,112 @@
+*/
+
+package prometheus
+
+import (
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Package prometheus sets the cache DefaultMetricsFactory to produce
+// prometheus metrics. To use this package, you just have to import it.
+
+func init() {
+	cache.SetReflectorMetricsProvider(prometheusMetricsProvider{})
+}
+
+type prometheusMetricsProvider struct{}
+
+func (_ prometheusMetricsProvider) NewListsMetric(name string) cache.CounterMetric {
+	lists := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "lists",
+		Help:      "Total number of API lists done by the reflector: " + name,
+	})
+	prometheus.Register(lists)
+	return lists
+}
+
+func (_ prometheusMetricsProvider) NewListDurationMetric(name string) cache.SummaryMetric {
+	listDuration := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "list_duration",
+		Help:      "How long an API list takes to return and decode for reflector: " + name,
+	})
+	prometheus.Register(listDuration)
+	return listDuration
+}
+
+func (_ prometheusMetricsProvider) NewItemsInListMetric(name string) cache.SummaryMetric {
+	itemsPerList := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "items_per_list",
+		Help:      "How many items an API list returns to reflector: " + name,
+	})
+	prometheus.Register(itemsPerList)
+	return itemsPerList
+}
+
+func (_ prometheusMetricsProvider) NewWatchesMetric(name string) cache.CounterMetric {
+	watches := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "watches",
+		Help:      "Total number of API watches done by the reflector: " + name,
+	})
+	prometheus.Register(watches)
+	return watches
+}
+
+func (_ prometheusMetricsProvider) NewShortWatchesMetric(name string) cache.CounterMetric {
+	watches := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "short_watches",
+		Help:      "Total number of short API watches done by the reflector: " + name,
+	})
+	prometheus.Register(watches)
+	return watches
+}
+
+func (_ prometheusMetricsProvider) NewWatchDurationMetric(name string) cache.SummaryMetric {
+	watchDuration := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "watch_duration",
+		Help:      "How long an API watch takes to return and decode for reflector: " + name,
+	})
+	prometheus.Register(watchDuration)
+	return watchDuration
+}
+
+func (_ prometheusMetricsProvider) NewItemsInWatchMetric(name string) cache.SummaryMetric {
+	itemsPerWatch := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "items_per_watch",
+		Help:      "How many items an API watch returns to reflector: " + name,
+	})
+	prometheus.Register(itemsPerWatch)
+	return itemsPerWatch
+}
+
+func (_ prometheusMetricsProvider) NewLastResourceVersionMetric(name string) cache.GaugeMetric {
+	rv := prometheus.NewGauge(prometheus.GaugeOpts{
```

Comment:
> Can the last resource version go down?

Maybe.  You can imagine cases like etcd backup restoration or a list against a latent etcd.